src/include/btree_cmp_inline.h

```diff
@@ -14,12 +14,182 @@
 #endif
 #endif
 
-#if defined(HAVE_ARM_NEON_INTRIN_H)
-#include <arm_neon.h>
-#endif
-/* 16B alignment */
-#define WT_ALIGNED_16(p) (((uintptr_t)(p)&0x0f) == 0)
 #define WT_VECTOR_SIZE 16 /* chunk size */
+#define WT_COMPARE_SHORT_MAXLEN 16
+
+#ifdef HAVE_X86INTRIN_H
+
+#define WT_VECTOR_MASK 0xFFFF
+/*
+ * __lex_compare_gt_16 --
+ *     Lexicographic comparison routine for data greater than 16 bytes. Returns: < 0 if user_item is
+ *     lexicographically < tree_item, = 0 if user_item is lexicographically = tree_item, > 0 if
+ *     user_item is lexicographically > tree_item. We use the names "user" and "tree" so it's clear
+ *     in the btree code which the application is looking at when we call its comparison function.
+ */
+static WT_INLINE int
+__lex_compare_gt_16(const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp)
+{
+    __m128i res_eq, t, u;
+    int32_t eq_bits;
+    size_t i, final_bytes;
+
```

Comment:
Consider adding a debug-only assert at the start of the `gt_16` and `le_16` function that the len is > or <= 16 respectively since the implementations assume it without checking. I know that they are supposed to be implementation details of functions that dispatch to the right one based on size, but since there is nothing that prevents other callers, it may be worth validating.

Reply:
added but it is bit ugly because we cannot pass the session to __wt_lex_compare.

Reply:
I reverted this